File: src/diffler/collectors/registry.py
"""Collector registry and template dependency analysis."""

from __future__ import annotations

import logging
import re
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from diffler.collectors.base import Collector

logger = logging.getLogger(__name__)

# Regexes to find Jinja2 variable references and function calls
_VAR_REF = re.compile(r"\{\{[\s\-]*([a-zA-Z_][a-zA-Z0-9_]*)\b")
_FUNC_CALL = re.compile(r"\{\{[\s\-]*([a-zA-Z_][a-zA-Z0-9_]*)\s*\(")
# Also scan {% ... %} tags for variable references (e.g. {% for repo in repositories %})
_TAG_VAR_REF = re.compile(r"\{%[\+\s\-]*(?:for\s+\w+\s+in|if|elif|set\s+\w+\s*=)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b")


class CollectorRegistry:
    """Manages registered collectors and determines which ones to run."""

    def __init__(self) -> None:
        self._collectors: dict[str, Collector] = {}

    def register(self, collector: Collector) -> None:
        """Register a collector instance."""
        self._collectors[collector.name] = collector
        logger.debug("Registered collector: %s", collector.name)

    def discover_needed(self, template_source: str) -> list[str]:
        """Analyze template source to find which collectors are needed.

        Scans the template for references to collector ``template_refs``
        and resolves dependencies.

        Args:
            template_source: Raw Jinja2 template text.

        Returns:
            Ordered list of collector names to execute.
        """
        refs = set()
        refs.update(_VAR_REF.findall(template_source))
        refs.update(_FUNC_CALL.findall(template_source))
        refs.update(_TAG_VAR_REF.findall(template_source))

        needed: set[str] = set()
        for name, collector in self._collectors.items():
            if collector.template_refs & refs:
                needed.add(name)

        # Resolve dependencies
        resolved: list[str] = []
        seen: set[str] = set()

        def _resolve(name: str) -> None:
            if name in seen or name not in self._collectors:
                return
            collector = self._collectors[name]
            for dep in collector.dependencies:
                _resolve(dep)
            seen.add(name)
            resolved.append(name)

        for name in list(needed):
            _resolve(name)

        logger.info("Collectors needed: %s", resolved)
        return resolved

    def run(self, names: list[str], ctx: "CollectorContext") -> dict[str, str]:
        """Execute the specified collectors in order.

        Returns:
            Dict mapping collector name to status: "success", "skipped",
            or "failed".
        """
        status: dict[str, str] = {}
        for name in names:
            collector = self._collectors.get(name)
            if not collector:
                continue
            if collector.optional and ctx.rate_limit_remaining < 500:
                logger.warning(
                    "Skipping optional collector %s (rate limit: %s)",
                    name,
                    ctx.rate_limit_remaining,
                )
                status[name] = "skipped"
                continue
            try:
                collector.run(ctx)
                status[name] = "success"
            except Exception:
                logger.exception("Collector %s failed", name)
                status[name] = "failed"
                if not collector.optional:
                    raise
        return status

File: packages/diffler/src/collectors/registry.ts
import type { Collector } from "./types.js";
import type { CollectorContext } from "./context.js";

// Regexes to find Jinja2/Nunjucks variable references and function calls
const VAR_REF = /\{\{[\s\-]*([a-zA-Z_][a-zA-Z0-9_]*)\b/g;
const FUNC_CALL = /\{\{[\s\-]*([a-zA-Z_][a-zA-Z0-9_]*)\s*\(/g;
// Also scan {% ... %} tags for variable references
const TAG_VAR_REF =
  /\{%[\+\s\-]*(?:for\s+\w+\s+in|if|elif|set\s+\w+\s*=)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b/g;

export class CollectorRegistry {
  private collectors = new Map<string, Collector>();

  register(collector: Collector): void {
    this.collectors.set(collector.name, collector);
  }

  discoverNeeded(templateSource: string): string[] {
    const refs = new Set<string>();

    let m: RegExpExecArray | null;

    VAR_REF.lastIndex = 0;
    while ((m = VAR_REF.exec(templateSource)) !== null) {
      refs.add(m[1]);
    }

    FUNC_CALL.lastIndex = 0;
    while ((m = FUNC_CALL.exec(templateSource)) !== null) {
      refs.add(m[1]);
    }

    TAG_VAR_REF.lastIndex = 0;
    while ((m = TAG_VAR_REF.exec(templateSource)) !== null) {
      refs.add(m[1]);
    }

    const needed = new Set<string>();
    for (const collector of this.collectors.values()) {
      for (const ref of collector.templateRefs) {
        if (refs.has(ref)) {
          needed.add(collector.name);
        }
      }
    }

    // Resolve dependencies
    const resolved: string[] = [];
    const seen = new Set<string>();

    const resolve = (name: string): void => {
      if (seen.has(name) || !this.collectors.has(name)) return;
      const collector = this.collectors.get(name)!;
      for (const dep of collector.dependencies) {
        resolve(dep);
      }
      seen.add(name);
      resolved.push(name);
    };

    for (const name of needed) {
      resolve(name);
    }

    return resolved;
  }

  async run(names: string[], ctx: CollectorContext): Promise<Record<string, string>> {
    const status: Record<string, string> = {};

    for (const name of names) {
      const collector = this.collectors.get(name);
      if (!collector) continue;

      if (collector.optional && ctx.rateLimitRemaining < 500) {
        console.warn(`Skipping optional collector ${name} (rate limit: ${ctx.rateLimitRemaining})`);
        status[name] = "skipped";
        continue;
      }

      try {
        const value = await collector.collect(ctx);
        ctx.set(name, value);
        status[name] = "success";
      } catch (err) {
        console.error(`Collector ${name} failed:`, err);
        status[name] = "failed";
        if (!collector.optional) {
          throw err;
        }
      }
    }

    return status;
  }
}
